### Executive Summary:

In this exercise, a greedy algorithm is examined by finding the lowest cost to employ six or less guards working 24 shifts at a hypothetical security company providing services for a construction building. The wage structure is guards 8 or less hours are paid $15 per hour with overtime hours being paid an additional 5 dollars an hour. Basically a greedy algorithm is an approximation algorithm where at each step it will pick the locally optimal solution, and at the end you are left with an outcome that is either close to or is the globally optimal solution. In other words, in a cost minimization problem, at each step it will choose to lowest cost option first until all the requirements have been satisfied. Greedy algorithms are mainly conceptual, and many algorithms are considered greedy such as Breadth-first search and Dijkstra's algorithm. It is useful to the data engineer to know how to come up with greedy algorithms especially when they need to find an approximate solution to an NP complete problem. Thus, in this exercise, the combination with the most cost effective solution that is produced by a greedy algorithm is evaluated. Further explanation on how the algorithm works, and its relationship to Big O notation is also discussed.

Based on the results of this exercise, it is recommended that a greedy algorithm is used for NP complete problems that need approximate solutions that are close to one that is the globally optimal solution. Though there are cases where a greedy algorithm may not be perfect or the best solution to a problem, such as a knapsack problem, it is an effective means of providing an adequate solution where issues can be time sensitive, and finding the global optimum becomes secondary.       

### Import Libraries:

In [1]:
from collections import deque
from timeit import default_timer as timer

### Define Function for Greedy Algorithm:
 - A greedy algorithm is an approximation algorithm. When finding the exact solution will take too much time, an approximation algorithm can be just as effective. Approximation algortihms are assessed by: 
     - How fast they are
     - How close they are to a globally optimum solution [1]
 - Greedy algorithms are very simple to come up with. At each step, they will pick the locally optimal solution until it ends with one that is close to or is the globally optimal solution.
 - It is not always the perfect solution, but its is an algorithm that solves a problem that is acceptable. It is especially useful when finding the exact global optimal becomes secondary due to time considerations, such as in processing systems that provide analytics on streams [2].    
 - Greedy algorithms are best used in cases where a solution is needed for an NP complete problem. This is useful for the data engineer to know, as there are many problems that are NP complete, where an exact solution would cost too much time and resources.
 - Big O notation is written as **O(*2^n*)** time for the problem in this exercise. Generally, greedy algorithms will still go through every possible subset which are *2^n*, and will pick the locally optimal solution. On the other had, in an NP complete problem, finding the exact solution takes **O(*n!*)** time. The difference is that if there are 100 subsets, an exact algorithm will approximately take **4x10^21** years to solve the problem, whereas a greedy algorithm will take approximately **16.67** minutes [1].
 - The greedy algorithm defined below is similar to the Breadth-first search algorithm. The algorithm works by:
     - First, creating a dictionary that includes the total cost for each guard combination and total cost per combination.
     - Second, utilizes a FIFO (First In, First Out) method where the dictionary is queued, and the lowest total cost is tracked and updated through a process of dequeuing. The algorithm basically chooses the locally optimal solution at each step, takes the first local optimal and makes it the globally optimal solution unless another locally optimal solution is found. 
 - Therefore, an algorithm similar to a Breadth-first search algorithm was used, as it is an efficient, effective, and fast running algorithm. It is also greedy.

In [2]:
# Function to find optimal solution
def search_optimal(shifts):
    # dictionary that includes total cost per unit and combination
    totals = {}
    keys = list(shifts.keys())
    items = list(shifts.items())
    for i in range(len(keys)):
        values = list(shifts.values())
        for e in range(len(values)):
            total_hrs_cost=[]
            for k in range(len(values[e])):
                if values[e][k] > 8:
                    over_cost = ((values[e][k]-8)*20) + ((values[e][k]-(values[e][k]-8))*15)
                    total_hrs_cost.append(over_cost)
                    if items[i] == (keys[i], values[e]):
                        totals[keys[i]] = values[e], total_hrs_cost
                else:
                    reg_cost = values[e][k]*15
                    total_hrs_cost.append(reg_cost)
                    if items[i] == (keys[i], values[e]):
                        totals[keys[i]] = values[e], total_hrs_cost

    # find optimal
    find_min = list(totals.items()) 
    min_search = deque()
    min_search += find_min
    least = None
    value = min_search.popleft()
    for i in range(0, len(find_min)):
        if sum(find_min[i][1][1]) < sum(value[1][1]):
            least = find_min[i]
            value = min_search.popleft()
        else:
            least = value
    return  [least, sum(least[1][1])] #return first combination with least amount of cost and total   

### Shift Combinations:

In [3]:
#Number of guards and working hours
guard_shifts = {'2 Guards':[12, 12], '3 Guards':[12, 8, 4], '4 Guards ':[8, 6, 6, 4],
                '5 Guards':[6, 6, 4, 4, 4], '6 Guards':[4, 4, 4, 4, 4, 4]}

### Optimal Solution:

In [12]:
# Optimal solution with execution time recorded
start = timer()
optimal_solution = search_optimal(guard_shifts)
end = timer()
exec_time = (end-start)*1000
print('Optimal Solution: \n', 'Guards:', optimal_solution[0][0], '\n Hours Each:', optimal_solution[0][1][0],
     '\n Total Cost:', optimal_solution[1], '\n Execution Time:', exec_time)

Optimal Solution: 
 Guards: 4 Guards  
 Hours Each: [8, 6, 6, 4] 
 Total Cost: 360 
 Execution Time: 0.12449999849195592


### Findings and Alternative Solution:
 - The local optimal solution in this case was four guards as that was the first solution with the least amount of cost, 360 dollars. The algorithm also found that this was the globally optimal solution. However, for this problem there are many alternative solutions as the problem was only to solve how many guards with the least amount of cost, where all guards working 8 or under 8 hours get paid the same. However, there are many solutions for 6 guards (about 720 combinations). Therefore, any of the other combinations specified above where the guards work 8 hours and under are just as effective, such as 5 guards or 6 guards. The problem would also be different if there was an overlap between hours, or if there were a minimum requirement of guards. However, the approximations and combinations would remain similar, as greedy algorithms are focused on finding locally optimal solutions. Again this is useful to the data engineer, as this was an example of an NP complete problem. Though the algorithm went through every subset, it only provided an approximation by using the locally optimal solution, and it was not focused on finding the exact solution which would have taken much more that 0.1245 milliseconds.    

### Conclusion:
In this exercise, a greedy algorithm was examined by finding the lowest cost to employ six or less guards working 24 shifts at a hypothetical security company providing services for a construction building. Basically a greedy algorithm is an approximation algorithm where at each step it will pick the locally optimal solution, and at the end you are left with an outcome that is either close to or is the globally optimal solution. Greedy algorithms are mainly conceptual, and many algorithms are considered greedy such as Breadth-first search and Dijkstra's algorithm. Thus, in this exercise, the combination with the most cost effective solution that was produced by a greedy algorithm was evaluated.

Big O notation for the greedy algorithm used in this exercise is written as **O(*2^n*)**. Generally, greedy algorithms will still go through every possible subset which are *2^n*, and will pick the locally optimal solution. On the other had, in an NP complete problem, finding the exact solution takes **O(*n!*)** time.

The greedy algorithm defined in this exercise was similar to the Breadth-first search algorithm. The algorithm worked by:
 - First, creating a dictionary that included the total cost for each guard combination and total cost per combination
 - Second, utilizee a FIFO (First In, First Out) method where the dictionary was queued, and the lowest total cost was tracked and updated through a process of dequeuing. The algorithm basically chose the locally optimal solution at each step, took the first local optimal and made it the globally optimal solution unless another locally optimal solution was found.

Therefore, an algorithm similar to a Breadth-first search algorithm was used, as it is an efficient, effective, and fast running algorithm. It is also greedy.

Based on the results of this exercise, it is recommended that a greedy algorithm is used for NP complete problems that need approximate solutions that are close to one that is the globally optimal solution. Though there are cases where a greedy algorithm may not be perfect or the best solution to a problem, such as a knapsack problem, it is an effective means of providing an adequate solution where issues can become time sensitive, and finding the global optimum becomes secondary, such as in processing systems that provide analytics on streams. It is useful to the data engineer to know how to come up with greedy algorithms especially when they need to find an approximate solution to an NP complete problem, where finding an exact solution would cost too much time and resources.

### References:
[1] Bhargava, A. Y.(2016.) Grokking Algorithms: An Illustrated Guide for Programmers and Other Curious People. Shelter Island, N.Y.: Manning.

[2] Kleppmann, M. (2017.) Designing Data-Intensive Applications: The Big Ideas behind Reliable, Scalable, and Maintainable Systems. Sebastopol, Calif.: O’Reilly.